tests/integration_tests/datasources/test_none.py

```diff
@@ -0,0 +1,64 @@
+"""NoCloud datasource integration tests."""
+import json
+
+import pytest
+from pycloudlib.lxd.instance import LXDInstance
+
+from tests.integration_tests.decorators import retry
+from tests.integration_tests.instances import IntegrationInstance
+from tests.integration_tests.integration_settings import PLATFORM
+from tests.integration_tests.util import verify_clean_log
+
+DS_NONE_BASE_CFG = """\
+datasource_list: [None]
+datasource:
+  None:
+    metadata:
+      instance-id: my-iid-uuid
+    userdata_raw: |
+      #cloud-config
+      runcmd:
+      - touch /var/tmp/success-with-datasource-none
+"""
+
+LXD_METADATA_DS_NONE_BASE_CFG = """\
+  /etc/cloud/cloud.cfg.d/99-force-dsnone.cfg:
+    when:
+    - create
+    - copy
+    create_only: false
+    template: dsnone.tpl
+    properties: {}
+"""
+
+
+@retry(tries=30, delay=1)
+def wait_for_cloud_init_status_file(instance: LXDInstance):
+    status_file = instance.read_from_file("/run/cloud-init/status.json")
+    assert len(status_file)
+
+
+@pytest.mark.lxd_use_exec
+@pytest.mark.skipif(
+    PLATFORM != "lxd_container",
+    reason="Requires LXD container with custom metadata setup",
+)
+def test_datasource_none_discovery(client: IntegrationInstance):
+    """Integration test for #4635.
+
+    Test that DataSourceNone detection (used by live installers) doesn't
+    generate errors or warnings.
+    """
+    log = client.read_from_file("/var/log/cloud-init.log")
+    verify_clean_log(log)
+    # Limit datasource detection to DataSourceNone.
+    client.write_to_file(
+        "/etc/cloud/cloud.cfg.d/99-force-dsnone.cfg", DS_NONE_BASE_CFG
+    )
+    client.execute("cloud-init clean --logs --reboot")
```

Comment:
I think we should do `client.execute("cloud-init clean --logs")` followed by `client.restart()` here, otherwise there's a possibility your `wait_for_cloud_init_status_file()` will return before the instance has even gone down.

Reply:
`wait_for_cloud_init_status_file` doesn't return unless /run/cloud-init/status.json is non-empty. We won't get to this case because cloud-init clean trashes the link target in `/v/l/c/instance/` that `/run/cloud-init/status.json` points to so any attempt to cat /run/cloud-init/status.json returns FileNotFound errors which exits non-zero. Our `read_from_file` raises an IOError in those cases which triggers a retry on `wait_for_cloud_init_status_file`.  The reason I want to avoid client.restart() is because it blocks until cloud-init completes, and in the failure mode here, cloud-init never completes due to unhandled Tracebacks which leave `cloud-init status --wait` blocking in a running state indefinitely. Since `client.restart()` in our integration tests doesn't provide a `wait=False` I constructed the wait_for_cloud_init_status_file as a workaround for this case to at least give us visibility to any warning/error conditions that show up even if cloud-init never completes.

Reply:
Ah, right. I wasn't thinking about clean removing the link. This is good